### PR TITLE
The big refactor PR

### PR DESCRIPTION
**Pull request by captivus**
This PR adds Poetry as the package and dependency manager for GPT Engineer!

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -39,64 +18,66 @@ classifiers = [
...
+typer = ">=0.3.2"
+rudder-sdk-python = ">=2.0.2"
+dataclasses-json = "0.5.7"
+tiktoken = ">=0.0.4"
+tabulate = "0.9.0"
+python-dotenv = ">=0.21.0"
+langchain = ">=0.0.335"
+llama-index = ">=0.8.49"
+rank-bm25 = ">=0.2.2"
+tree_sitter_languages = ">=1.8.0"
```
Comment (ATheorell) [inline pyproject.toml:39]: Can you put these in a new list under [tool.poetry.extras] called something like experimental + relevant readme update?

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -22,13 +22,29 @@ jobs:
       - uses: actions/setup-python@v4
         with:
           python-version: "3.10"
-          cache: pip
```
Comment (ErikBjare) [inline .github/workflows/ci.yaml:25]: This supports `poetry`!

**Comment by ATheorell** [comment]
One test is failing (self-heal using cli_agent), because a key is not found in the prompt cache in CachingAI. The key is found when I run it locally. The cache is in the git-tracked file tests/ai_cache.json

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -22,13 +22,29 @@ jobs:
...
 
-      - name: Install package
-        run: pip install -e .
+      # Install Poetry
+      - name: Install Poetry
+        run: |
+          curl -sSL https://install.python-poetry.org | python3 -
```
Comment (ErikBjare) [inline .github/workflows/ci.yaml:29]:
```suggestion
          pipx install poetry
```

**Comment by captivus** [comment]
I get an error when I choose "n" to not run the code after generation.

```
Do you want to execute this code? (Y/n)

python3 -m venv venv
source venv/bin/activate
pip install -r requirements.txt
python main.py


n
Ok, not executing the code.
Traceback (most recent call last):

  File "/home/captivus/projects/gpt-engineer/gpt-engineer/.venv/bin/ge", line 6, in <module>
    sys.exit(app())
             ^^^^^

  File "/home/captivus/projects/gpt-engineer/gpt-engineer/gpt_engineer/applications/cli/main.py", line 217, in main
    store.upload(files_dict)

  File "/home/captivus/projects/gpt-engineer/gpt-engineer/gpt_engineer/core/default/file_store.py", line 17, in upload
    for name, content in files.items():
                         ^^^^^^^^^^^
```

@ATheorell

**Comment by ATheorell** [comment]
Fixed @captivus

**Pull request merged**